coordinator/src/escrow_coordinator/mod.rs

```diff
@@ -29,52 +31,66 @@ impl EscrowCoordinator {
-                if let Some(decrypted) = self
-                    .nostr_client
-                    .decrypt_msg(&event.content, &event.author())
-                    .await
-                {
-                    dbg!("Received event: {:?}", &decrypted);
-                    if let Ok((contract_hash, contract)) = self.parse(decrypted.as_str()).await {
-                        if self.pending_contracts.contains_key(&contract_hash) {
-                            self.pending_contracts.remove(&contract_hash);
-                            self.begin_trade(&contract_hash, &contract).await?;
-                        } else {
-                            self.pending_contracts.insert(contract_hash, contract);
+        loop {
+            match notifications.recv().await {
+                Ok(notification) => {
+                    if let RelayPoolNotification::Event { event, .. } = notification {
+                        if let Ok(unwrapped_gift) =
+                            self.nostr_client.client.unwrap_gift_wrap(&event).await
+                        {
+                            let rumor = unwrapped_gift.rumor;
+                            if rumor.kind == Kind::PrivateDirectMessage {
+                                if let Ok((contract_hash, contract)) =
+                                    EscrowCoordinator::parse_contract(&rumor.content)
+                                {
+                                    dbg!("Received contract: {}", &contract.trade_description);
+                                    if self.pending_contracts.contains_key(&contract_hash) {
+                                        self.pending_contracts.remove(&contract_hash);
+                                        let _ = self
+                                            .begin_trade(&contract_hash, &contract)
+                                            .await
+                                            .inspect_err(|e| {
+                                                //todo: use logger instead
+                                                println!(
+                                                    "Got error while beginning a trade: {}",
+                                                    e
+                                                );
+                                            });
+                                    } else {
+                                        self.pending_contracts.insert(contract_hash, contract);
+                                    }
+                                }
+                            }
                         }
+                    } else if RelayPoolNotification::Shutdown == notification {
+                        break Err(anyhow!(
+                            "Got shutdown notification, breaking coordinator lopp!"
+                        ));
                     }
                 }
+                Err(RecvError::Closed) => {
```

Comment:
As the coordinator should be a long running application maybe we could also try to reinitialize the coordinator (close all connection/subscriptions and set them up again) on error, and check if the error doesn't happen again in a specific timeframe (e.g. 30 mins), so it would be more reliable and not crash if there is a cause (e.g. short internet outage) that closes the relay pool from time to time. 
I guess if it runs in a docker container with restart: always this would also be handled, but seems like a cleaner solution to implement some recovery logic in the coordinator itself.

Reply:
Addressed separately in Issue https://github.com/f321x/cashu-escrow-kit/issues/10

Reply:
Handled in separate issue https://github.com/f321x/cashu-escrow-kit/issues/11